### PR TITLE
[WIP] Improve Developing.md

### PR DESCRIPTION
**Pull request by pierregoutagny**
Task: #440 

Signed-off-by: Pierre GOUTAGNY <pierregoutagny@gmail.com>

### CODE REVIEW

**Comment by akosyakov** [comment]
@pierregoutagny looking good, could you sign off your commits

For existings commits use `git rebase -i` and then reword commits to add  `Signed-off-by: Joe Smith <joe.smith@email.com>` at the end of each commit message, with your name and email.
For new commits do `git commit -s`.

**Comment by hexa00** [comment]
Also you don't need to last merge commit

**Comment by akosyakov** [comment]
@pierregoutagny when you rebase with -i, you have to force push to overwrite existing git history otherwise you add new commits instead of signing off existing

What you can do now, it is again `git rebase -i origin/master` and then you mark commits except signed with `drop` and then you do `git push -f origin HEAD` to overwrite git history of your PR.

Sorry for the inconvenience...

**Comment by pierregoutagny** [comment]
Well, thank you for your help! I'm sorry I forgot the `-s` in the first place.

**Comment by hexa00** [comment]
That looks much better :)

You want to do the change for --unsafe-perm in this PR or leave it like that?

**Comment by pierregoutagny** [comment]
I was thinking about fixing #440 point by point in this PR, but if you recommend I split it into different PRs I can do so.

**Comment by hexa00** [comment]
It's fine in this PR :) thx!

**Comment by akosyakov** [comment]
One can see the resulted document here: https://github.com/pierregoutagny/theia/blob/improve-developing-md/doc/Developing.md

**Comment by akosyakov** [comment]
I would like to rename `tl;dr` to `Quick Start`. Objections?

**Comment by kittaakos** [comment]
tldr means exactly what it has to. But if others prefer quick start, then quick start.

**Comment by akosyakov** [comment]
@kittaakos for me it is a language barrier if you don't know you are lost. I want people to get started not google what `tl;dr` means.

**Comment by kittaakos** [comment]
A couple of other comments:
 - poor Windows developers have almost zero support :(
 - VSCode is a brand too, I would consider a blog post a documentation better if things are written correctly: VSCode and not vscode, JavaScript not Javascript Node.js and not nodejs and so on.

**Comment by akosyakov** [comment]
@pierregoutagny `Table of Contents` header could be removed, it is ok if the index belongs to the end of `How to build Theia and the example applications` section

**Comment by akosyakov** [comment]
@kittaakos let's tackle windows separately, don't want @pierregoutagny get overwhelmed :) added the point about proper brand names to the issue

**Comment by akosyakov** [comment]
@pierregoutagny are you going to work on remaining points?

**Comment by pierregoutagny** [comment]
Yeah I'll continue this afternoon

File: doc/Developing.md [hunk 1]
```diff
@@ -19,6 +42,27 @@ So far it has been confirmed that there are some problems and errors using nodej
...
 [nvm](https://github.com/creationix/nvm) is recommended to easily switch between node versions.
 
+## Quick Start
+
+To build and run the browser example:
+
+    git clone https://github.com/theia-ide/theia \
+    && cd theia \
+    && yarn \
+    && cd examples/browser \
```
Comment (kittaakos) [inline doc/Developing.md]: Question: shouldn't we add `yarn run rebuild:browser` here? Just to be able to switch between electron and the browser versions.

Comment (kittaakos) [reply doc/Developing.md]: I meant one above, sorry.

File: doc/Developing.md [hunk 2]
```diff
@@ -211,3 +234,11 @@ Theia uses native modules and also requires Python 2.x to be installed on the sy
 
 [all-in-one packages]: https://github.com/felixrieseberg/windows-build-tools
 [bug]: https://github.com/theia-ide/theia/issues
+
+### Root privileges errors
+When trying to install with root privileges, you might encounter errors such as `cannot run in wd`.
```
Comment (kittaakos) [inline doc/Developing.md]: Can we reference this closed issue? https://github.com/theia-ide/theia/issues/266#issuecomment-313328699

File: doc/Developing.md (hunk 1 above)
Comment (akosyakov) [reply doc/Developing.md]: One does not need to run it to start the browser example after cloning. If one wants to run not from the scratch there is a section covering it: https://github.com/pierregoutagny/theia/blob/improve-developing-md/doc/Developing.md#run-the-browser-based-example-application

File: doc/Developing.md [hunk 3]
```diff
@@ -11,6 +10,30 @@ The following instructions are for Linux and macOS.
...
 
+ * [**Prerequisites**](#prerequisites)
+ * [**Quick Start**](#quick-start)
+ * [**Clone the repository**](#clone-the-repository)
+ * [**Build core, extensions and examples packages**](#build-core-extensions-and-examples-packages)
+ * [**Run the browser-based example application**](#run-the-browser-based-example-application)
+ * [**Run the electron-based example application**](#run-the-electron-based-example-application)
+ * [**Rebuilding**](#rebuilding)
+ * [**Watching**](#watching)
+	 * [core and extension packages](#core-and-extension-packages)
```
Comment (kittaakos) [inline doc/Developing.md]: Just to be consistent: the core and the extension packages. Do you agree? If so, can we have that?

File: doc/Developing.md (hunk 1 above)
Comment (kittaakos) [reply doc/Developing.md]: True. Does the `rebuild:electron` mandatory when building the electron example from a fresh clone? I am just asking because it seems to be inconsistent somehow.

Comment (akosyakov) [reply doc/Developing.md]: yes, for electron is mandatory

**Comment by pierregoutagny** [comment]
I'm working on #440 point 4:
I was thinking about adding the "repo structure" section **just before the "Rebuilding section"** so that once you've understood the structure you can start making changes and then rebuild. However, it might as well be **a sub-section of the existing "Cloning the repository" section**, answering the question "Well, I just cloned a repo, but how is it actually structured?".

What do you feel would be the most consistent?

**Comment by akosyakov** [comment]
@pierregoutagny I prefer the second option, e.g. `The repository structure` after `Cloning the repository`, so the reader knows what the extension or example package mean, before moving to building them.

**Comment by pierregoutagny** [comment]
Also, is it OK if I copy and edit a bit the list in https://github.com/theia-ide/theia/issues/359#issuecomment-319008318 ? You actually explained everything very well.

**Comment by akosyakov** [comment]
@pierregoutagny it is ok if it makes sense to others, especially to you since you have less implicit assumptions :) so if something is implicit you can explain it or ask here to clarify

File: doc/Developing.md [hunk 4]
```diff
@@ -73,6 +73,16 @@ variable in your shell:
 
     THEIA=$PWD/theia
 
+## Repository structure
+
+Theia repository has multiple folders:
```
Comment (pierregoutagny) [inline doc/Developing.md]: I chose the word "folder" to avoid confusion with the folder named `packages`

**Comment by pierregoutagny** [comment]
I'm not sure about what the `scripts` folder actually does... Is it a **build** scripts folder?
Please tell me if there are other things I didn't understand.

**Comment by akosyakov** [comment]
`scripts` folder contains js scripts used by npm scripts

File: doc/Developing.md (hunk 4 above)
Comment (akosyakov) [reply doc/Developing.md]: the section looks good to me, I've added it to the index

**Comment by akosyakov** [comment]
@pierregoutagny congrats with your first PR and thank you!

**Comment by pierregoutagny** [comment]
Thank you a lot for helping me and welcoming me for my first PR!
